rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,627 @@
+
+    *c = newClient
+}
+```
+
+It will be the responsibility of the caller of `Refresh` to reinitialize any
+watches or streams that the client may have been running prior to the refresh.
+
+### Certificate Specification
+
+The bot will be configured with one or more certificate specs, which can be
+supplied on the command line in the format of:
+
+```
+--cert=DESTINATION,RELOAD
+```
+
+...or via a `tbot.yaml` configuration file.
+
+*TODO*: how can a user specify DNS names to include in the certs?
+
+*TODO* (Tim): Non-filesystem destinations are likely to be very difficult to
+encode as CLI flags (e.g. k8s secrets). Perhaps the CLI should only support
+local FS and other sinks should defer to `tbot.yaml`?
+
+#### Artifacts
+
+For internal use, the bot generates a set of renewable TLS certificates.
+This set of credentials is only used certificate renewal, user certificate
+generation, and establishing CA rotation watches. These artifacts should be
+persisted to some private location to allow the bot to recover after a process
+restart.
+
+| Kind                   | Path         | Notes                            |
+|------------------------|--------------|----------------------------------|
+| TLS CA certificates    | `tlscacerts` | Teleport cluster CA certs (both) |
+| TLS client certificate | `tlscert`    | Signed by Teleport `UserCA`      |
+| Private key (TLS)      | `key`        |                                  |
+
+The bot then generates a set of non-renewable credentials for each configured
+credential sink. By default, the following artifacts are always generated:
+
+| Kind                      | Path          | Notes
+| -----------------------|---------------|-------
+| Private key            | `key`         | Shared for both SSH and TLS uses        |
+| SSH public key         | `key.pub`     | Required for OpenSSH compatibility      |
+| SSH client certificate | `sshcert`     | Signed by Teleport `UserCA`             |
+| SSH CA certificates    | `sshcacerts`  | TODO: needed?                           |
+| SSH known hosts        | `known_hosts` | Teleport `HostCA` certs, OpenSSH format |
+| TLS CA certificates    | `tlscacerts`  | Teleport CA certificates (both)         |
+| TLS client certificate | `tlscert`     | Signed by Teleport `UserCA`             |
+
+Notes:
+ * Users may optionally decide to disable persistence of the renewable
+   certificate. This has security benefits but prevents the bot from recovering
+   if stopped, as the renewable credentials would only exist in memory.
+ * TLS CA certificates are verified at first connect using CA pins, following
+   Teleport's usual node joining procedure.
+ * Unneeded certificate types/formats could be optionally disabled in
+   `tbot.yaml` config if desired.
+ * Potential in-memory certificate sinks should be evaluated for end-users
+   (Webhooks?)
+ * (TODO) Additional artifacts may be written if we decide to write app config
+   files at this stage.
+
+#### Destination
```

Comment:
It looks like you've been preferring the term "sink" over destination. I would just pick one and be consistent.

Reply:
I think they're both still distinct and important:
 * A destination contains files in a backend-agnostic way (with some configuration)
 * A sink is a destination + certificates and their configuration (which certificate types, config files to generate, etc)

I think the distinction is important because we'll also store the bot's own data (the renewable cert and other state) in a destination, and the configuration for this will share that configuration syntax. For example, I could see Kubernetes secrets as a useful backend both for the renewable cert and end-user certs. (Or no-op to keep it only in memory, etc)

That's not to say I haven't been using the two terms somewhat interchangeably, I'll see about consolidating terminology where it makes sense.